moveit_experimental/jog_arm/src/jog_arm/jog_arm_server.cpp

```diff
@@ -0,0 +1,1254 @@
+
+      collision_rate.sleep();
+    }
+  }
+}
+
+// Constructor for the class that handles jogging calculations
+JogCalcs::JogCalcs(const JogArmParameters& parameters, jog_arm_shared& shared_variables,
+                   const std::unique_ptr<robot_model_loader::RobotModelLoader>& model_loader_ptr)
+  : move_group_(parameters.move_group_name), tf_listener_(tf_buffer_)
+{
+  parameters_ = parameters;
+
+  // Publish collision status
+  warning_pub_ = nh_.advertise<std_msgs::Bool>(parameters_.warning_topic, 1);
+
+  // MoveIt Setup
+  // Wait for model_loader_ptr to be non-null.
+  while (ros::ok() && !model_loader_ptr)
+  {
+    ROS_WARN_THROTTLE_NAMED(5, NODE_NAME, "Waiting for a non-null robot_model_loader pointer");
+    ros::Duration(0.1).sleep();
+  }
+  const robot_model::RobotModelPtr& kinematic_model = model_loader_ptr->getModel();
+  kinematic_state_ = std::make_shared<robot_state::RobotState>(kinematic_model);
+  kinematic_state_->setToDefaultValues();
+
+  joint_model_group_ = kinematic_model->getJointModelGroup(parameters_.move_group_name);
+
+  std::vector<double> dummy_joint_values;
+  kinematic_state_->copyJointGroupPositions(joint_model_group_, dummy_joint_values);
+
+  // Wait for initial messages
+  ROS_INFO_NAMED(NODE_NAME, "Waiting for first joint msg.");
+  ros::topic::waitForMessage<sensor_msgs::JointState>(parameters_.joint_topic);
+  ROS_INFO_NAMED(NODE_NAME, "Received first joint msg.");
+
+  ROS_INFO_NAMED(NODE_NAME, "Waiting for first command msg.");
+  ros::topic::waitForMessage<geometry_msgs::TwistStamped>(parameters_.cartesian_command_in_topic);
+  ROS_INFO_NAMED(NODE_NAME, "Received first command msg.");
+
+  resetVelocityFilters();
+
+  jt_state_.name = move_group_.getJointNames();
+  jt_state_.position.resize(jt_state_.name.size());
+  jt_state_.velocity.resize(jt_state_.name.size());
+  jt_state_.effort.resize(jt_state_.name.size());
+
+  // Low-pass filters for the joint positions & velocities
+  for (size_t i = 0; i < jt_state_.name.size(); ++i)
+  {
+    velocity_filters_.emplace_back(parameters_.low_pass_filter_coeff);
+    position_filters_.emplace_back(parameters_.low_pass_filter_coeff);
+  }
+
+  // Initialize the position filters to initial robot joints
+  while (!updateJoints() && ros::ok())
+  {
+    pthread_mutex_lock(&shared_variables.joints_mutex);
+    incoming_jts_ = shared_variables.joints;
+    pthread_mutex_unlock(&shared_variables.joints_mutex);
+    ros::Duration(0.001).sleep();
+  }
+  for (std::size_t i = 0; i < jt_state_.name.size(); ++i)
+    position_filters_[i].reset(jt_state_.position[i]);
+
+  // Wait for the first jogging cmd.
+  // Store it in a class member for further calcs.
+  // Then free up the shared variable again.
+  geometry_msgs::TwistStamped cartesian_deltas;
+  moveit_experimental::JogJoint joint_deltas;
+  while (ros::ok() && (cartesian_deltas.header.stamp == ros::Time(0.)) && (joint_deltas.header.stamp == ros::Time(0.)))
+  {
+    ros::Duration(0.05).sleep();
```

Comment:
sometimes your while loops wait for 0.001 and here its 0.05. this should be a standard constant variable declared above